Adding cond and filterOrElse

blast-hardcheese opened the pull request:
I'd like to be able to conveniently convert an `A` into an `E`, providing either a static `E` or converting the `A` into an `E`.

- *Note:* `cond` could alternately be called `filter`/`filterOr`. I have not used ZIO enough to have a conception of the idiomatic naming scheme.

Inspiration:
#### `cond(A => Boolean, A => E1)`
- Cats: [`Validated.cond`](https://github.com/typelevel/cats/blob/76dbddf985d5b8ab556db0f13c01bc3062b7a60a/core/src/main/scala/cats/data/Validated.scala#L614)

#### `filterOrElse(A => Boolean, => E1)`
- Scala stdlib: [`Either.filterOrElse`](https://www.scala-lang.org/api/2.12.3/scala/util/Either.html#filterOrElse[A1%3E:A](p:B=%3EBoolean,zero:=%3EA1):scala.util.Either[A1,B])
- Scalaz: [`\/.filter`](https://github.com/scalaz/scalaz/blob/9ef08f7d2edc1286da2bdd4b1a55d201d80de4ca/core/src/main/scala/scalaz/Either.scala#L152)

dcsobral commented:
I think it would be preferable to take a `PartialFunction`. Like this:
```scala
  def rejectIf[R, E1, E2 <: E1, A](self: ZIO[R, E2, A])(pf: PartialFunction[A, E1]): ZIO[R, E1, A] = self.flatMap {
    case v if pf.isDefinedAt(v) => ZIO.fail(pf(v))
    case _ => self
  }
```

blast-hardcheese commented:
@dcsobral Good catch! Updating...

blast-hardcheese commented:
Oh, actually, I didn't see the `PartialFunction`, I was more focused on `flatMap(_ => self)` instead of `ZIO.success(...)`. I don't like `PartialFunction` here, as:
```
<console>:32: error: type mismatch;
 found   : Int => Long
 required: PartialFunction[Int,Long]
       foo(_.toLong)
             ^
```

dcsobral commented:
Using a partial function you get both methods, and with nice syntax.
```scala
def testCond = {
    val goodCase = unsafeRun(
      ZIO.succeed(0).rejectIf{ case a if a != 0 => s"$a was not 0"}.sandbox.either
    ) must_=== Right(0)

    val badCase = unsafeRun(
      ZIO.succeed(5).rejectIf{ case a if a != 0 => s"$a was not 0"}.sandbox.either
    ).left.map(_.failureOrCause) must_=== Left(Left("5 was not 0"))

    goodCase and badCase
  }

  def testFilterOrElse = { // Note I inverted 0 and 5, to make for a simpler case condition
    val goodCase = unsafeRun(
      ZIO.succeed(5).rejectIf{ case 0 => "Predicate failed!"}.sandbox.either
    ) must_=== Right(5)

    val badCase = unsafeRun(
      ZIO.succeed(0).rejectIf{ case 0 => "Predicate failed!"}.sandbox.either
    ).left.map(_.failureOrCause) must_=== Left(Left("Predicate failed!"))

    goodCase and badCase
  }
```

dcsobral commented:
It is true that you can't trivially convert a function into a partial function, but I still think it's better because a partial function is _exactly_ what this is. And `cond` is just a pf that ignores the input parameter. 

I'm not a fan of "`cond`" as a name. It's too generic a name. Not a fan of "`filterOrElse`" -- it is better, but `orElse` goes from error to no error, so `filterOrElse` doing the opposite might be unintuitive. That's why I went with `rejectIf`.

blast-hardcheese commented:
@dcsobral I'm onboard with `rejectIf`, though that would be `A => Boolean, => E1`. A similar name would need to exist to describe `A => Boolean, A => E1`.

I don't consider `{ case a => f(a) }` an acceptable answer for the base case, though `PartialFunction[A, E1]` is exactly what I'm trying to do. `rejectWhen` may be an acceptable answer as well, but rejecting the decomposed form of `cond(_.foo) _` is undesirable, imo.

jdegoes commented:
The generalization of this:

```scala
def cond(A => Boolean, A => E1)
```

is this:

```scala
trait ZIO[R, E, A] {
   def whenM(A => Boolean)(f: A => ZIO[R, E, A]): ZIO[R, E, A]

   def whenNotM(A => Boolean)(f: A => ZIO[R, E, A]): ZIO[R, E, A]
}
```

which also suggests variants:

```scala
trait ZIO[R, E, A] {
   def whenM_(A => Boolean)(f: ZIO[R, E, A]): ZIO[R, E, A]

   def whenNotM_(A => Boolean)(f: ZIO[R, E, A]): ZIO[R, E, A]
}
```

One could further specialize, but they are already relatively easy, e.g.:

```scala
foo.whenM_(_ < 0)(ZIO.fail("Must be positive"!))
```

Note there is existing ZIO precedent for the name `when` as well as the underscore for "unused" parameter variants.

jdegoes commented:
@dcsobral 

I like this combinator. I would tweak it a bit (ignoring variance):

```scala
sealed trait ZIO[R, E, A] {
 def collect[B](e: E)(pf: PartialFunction[A, B]): ZIO[R, E, B] = ???
 def collectM[B](e: E)(pf: PartialFunction[A, ZIO[R, E, B]]): ZIO[R, E, B] = ???
 def reject(pf: PartialFunction[A, E]): ZIO[R, E, A] = ???
 def rejectM(pf: PartialFunction[A, ZIO[R, E, E]]): ZIO[R, E, A] = ???
}
```

I'd definitely take these, although I think of them as separate from the `when*` combinators.

blast-hardcheese commented:
So, to clarify, I should update this PR to provide definitions for:
```scala
trait ZIO[R, E, A] {
  def collect[E1 <: E, B](e: E1)(pf: PartialFunction[A, B]): ZIO[R, E1, B] = ???
  def collectM[E1 <: E, B](e: E1)(pf: PartialFunction[A, ZIO[R, E, B]]): ZIO[R, E1, B] = ???
  def reject[E1 <: E](pf: PartialFunction[A, E1]): ZIO[R, E1, A] = ???
  def rejectM[E1 <: E](pf: PartialFunction[A, ZIO[R, E1, E1]]): ZIO[R, E1, A] = ???

  def whenM[E1 <: E](A => Boolean)(f: A => ZIO[R, E1, A]): ZIO[R, E1, A]
  def whenNotM[E1 <: E](A => Boolean)(f: A => ZIO[R, E1, A]): ZIO[R, E1, A]
  def whenM_[E1 <: E](A => Boolean)(f: ZIO[R, E1, A]): ZIO[R, E1, A]
  def whenNotM_[E1 <: E](A => Boolean)(f: ZIO[R, E1, A]): ZIO[R, E1, A]
}
```
?

Also, should the constraint on `whenM_` and `whenNotM_`'s `f` be `ZIO[R, E1, Any]` to explicitly indicate that it is not considered in the resultant `A`?

I'm most interested in the `A => Boolean` and `A => E1` variants, as I hardly use `PartialFunction` and I'd rather not have to `whenM(_.isFoo)(ZIO.fail(_))`, as it permits sideways `a => ZIO.pure(a)` which seems like an overloaded `flatMap`. At least the method name is comprehensible, so I'd welcome the extra finger strain in order to get closer to my end goal.

jdegoes commented:
> Also, should the constraint on whenM_ and whenNotM_'s f be ZIO[R, E1, Any] to explicitly indicate that it is not considered in the resultant A?

It would be used, because the implementation will be something like:

```scala
... = self.flatMap(a => if (p(a)) f(a) else ZIO.succeed(a))
```

> whenM(_.isFoo)(ZIO.fail(_)), as it permits sideways a => ZIO.pure(a) which seems like an overloaded 

It corresponds to changing the `A` value when a predicate holds.

> so I'd welcome the extra finger strain in order to get closer to my end goal.

One layer of additional specialization would yield:

```scala
sealed trait ZIO[R, E, A] {
  def whenFail(p: A => Boolean)(f: A => E): ZIO[R, E, A]
  def whenFail_(p: A => Boolean)(f: A => E): ZIO[R, E, A]
}
```

And upon further reflection, `when` is used in slightly different context in ZIO / Ruby / etc., so I think different names are warranted, and your original stdlib suggestion seems fine.

So perhaps something more like:

```scala
def filterOrElse(p: A => Boolean)(f: A => ZIO[R, E, A]): ZIO[R, E, A]
def filterOrElse_(p: A => Boolean)(f: ZIO[R, E, A]): ZIO[R, E, A]
```

I dropped the `M` because the non-monadic versions cannot be written, so there are no variants. And I omitted the `Not` variants (`filterNot`) because I think they're uncommon.

So in total you'd have:

```scala
def filterOrElse(p: A => Boolean)(f: A => ZIO[R, E, A]): ZIO[R, E, A] = ???
def filterOrElse_(p: A => Boolean)(zio: => ZIO[R, E, A]): ZIO[R, E, A] = filterOrElse(p)(_ => zio)

de filterOrFail(p: A => Boolean)(e: => E): ZIO[R, E, A] = filterOrElse_(p)(ZIO.fail(e))
```

Separately, I do think `collect` and `reject` would be quite nice, though that could be a separate ticket / PR if desired (although I'd happily take them in this one 😄).

blast-hardcheese commented:
Excellent. Thanks for the guidance!

blast-hardcheese commented:
OK. I've now implemented:

I just had one question, should `rejectM` actually be:
```
def rejectM[R1 <: R, E1 >: E, E2 <: E1](pf: PartialFunction[A, ZIO[R1, E2, E1]]): ZIO[R1, E1, A]
```
or some other variance? (Related to @dcsobral's [comment from earlier](https://github.com/zio/zio/pull/968#issuecomment-499293585))

Having `PF[A, ZIO[R1, E1, E1]]` seemed overly restrictive, but I'm not sure which way the variance makes the most sense.

There aren't many other methods that use both `E1` and `E2`.

jdegoes commented:
> Having PF[A, ZIO[R1, E1, E1]] seemed overly restrictive, but I'm not sure which way the variance makes the most sense.

Only `E1` occurs in the output of the method, whereas your proposed `E2` occurs in an input, in the `ZIO` effect which is already covariant in that type parameter.

Therefore, as far as I can tell, they _should_ be equivalent, although you could test that to make sure Scalac isn't being weird. 😄

blast-hardcheese commented:
While messing around, I also discovered that

```scala
self.flatMap {
  case v if ! p(v) => ???
  case _ => self
}
```
executes `self` twice, which is very much not what we want to do in that case. Adding that as a test as well.

iravid commented on core/shared/src/main/scala/zio/ZIO.scala:
```diff
@@ -1156,6 +1156,62 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
+   */
+  final def collectM[R1 <: R, E1 >: E, B](e: E1)(pf: PartialFunction[A, ZIO[R1, E1, B]]): ZIO[R1, E1, B] =
+    self.flatMap {
+      case v if pf.isDefinedAt(v) => pf(v)
```
Would it be possible to reformulate this (and `rejectM` below) using `PartialFunction#applyOrElse`? That would prevent double-evaluation of `pro.

dcsobral replied:
Ah, yes, `applyOrElse`! I knew there was something added to avoid double evaluation, but I completely forgot what it was.

jdegoes commented:
@blast-hardcheese 

Right, sorry about that, it should be:

```scala
self.flatMap {
  case v if ! p(v) => ???
  case v => ZIO.succeed(v)
}
```

blast-hardcheese commented:
OK, I believe I've addressed all concerns. I had to use `andThen` to turn `ZIO[R, E, E]` into `ZIO[R, E, A]` before `applyOrElse`ing so the result types would line up, other than that this seems to satisfy the constraints.

blast-hardcheese commented:
Of course! That one slipped right by me. I've addressed it now.

blast-hardcheese commented:
Pushed up one more change that addresses `pf.andThen(_.flatMap(...)).applyOrElse(..., ...)` being all on one line. I thought the formatting tool would have complained or fixed it if I split it onto two lines, but apparently that's permitted.

jdegoes commented:
@blast-hardcheese It seems Circle CI has stalled. Can you open a new pull request so we can get this passing? Thank you!

iravid commented:
Rebasing this on master would get things rolling CI-wise.

blast-hardcheese commented:
Updated and squashed commits, 🤞

jdegoes commented on core/shared/src/main/scala/zio/ZIO.scala:
```diff
@@ -1168,6 +1168,61 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
+  /**
+   * Fails with `e` if the predicate fails.
+   */
+  final def filterOrFail[E1 >: E](p: A => Boolean)(e: => E1): ZIO[R, E1, A] =
```
The only issue I see is with the parameter ordering for `filterOrElse_` and `filterOrElse`. Ordinarily, the error would be first, because it is most likely to be constant, while the predicate is most likely to vary. 

Is there a reason you have the current ordering?

blast-hardcheese replied:
Entirely matching the parameter ordering of the Scala stdlib as well as cats. I'm opening to swapping them if you'd prefer.

jdegoes replied:
Works for me!

jdegoes commented:
![giphy](https://user-images.githubusercontent.com/156745/59574565-b5673200-9085-11e9-937c-737e17d036ce.gif)

Thank you for this contribution, Devon! And congratulations on your first ZIO pull request... hope it won't be the last. 😉
